Return the final-theme value for a page

trm217 opened the pull request:
Adds a new `final-theme` value to the theme-context. (Closes #70)
The new value is equal to the `forced-theme`, if one is defined.
Otherwise it is set equal to the resolved theme.

pacocoursey commented:
I'm worried the API is getting too confusing with all these return values 😅 Will need to think about how to simplify it...

trm217 commented:
I agree with you.
In my opinion, it would make sense to return the forced theme in the resolvedTheme value directly. The naming would be fitting for what it does.

trm217 commented:
On a second thought, the forcedTheme prop seems kind of obsolete when the value is returned in the resolvedTheme.
I think it might be worth a thought to consider handling a forcedTheme like the 'system' theme.

Consider the following: `<ThemeProvider forcedTheme="light />"`. If we would allow 'forced' to be a treated as a theme, we could return 'forced' when accessing the theme attribute and if we want to access the actual forced-theme we could simply use the resolvedTheme attribute. With this the forcedTheme attribute would no longer be needed and we could simplify the api.
What do you think? @pacocoursey

pacocoursey commented:
I actually like that!


```ts
if (theme === 'system') {
  resolvedTheme: 'light' | 'dark'
}

if (theme === 'forced') {
  resolvedTheme: keyof typeof themes
}
```

trm217 commented:
I updated the code to now return theme=`forced` when a forcedTheme is set. I also updated the docs and the test.
I am unsure if the `Why is resolved-theme necessary` section also has to be updated or not.

pacocoursey commented on src/index.tsx:
```diff
@@ -65,6 +69,8 @@ export const ThemeProvider: React.FC<ThemeProviderProps> = ({
   const setTheme = useCallback(
     theme => {
+      // When a theme is forced it should not be possible to override it.
+      if (forcedTheme) return
```
Here's a tricky bug that I missed in 0.1.0 (and then fixed in 0.1.1):

1. Set theme to dark
2. Navigate to dark-forced page
3. In another tab, set theme to light (fires localStorage event)
4. The dark forced page receives the event (theme is now light)
4. Navigate back to non-forced page
5. Since we ignored the `setTheme` call, the theme is still considered forced/dark

pacocoursey replied:
I think we'll need an extra piece of state (or maybe just a `ref`) to keep track of a "pending"/"upcoming" theme, so that when forcedTheme goes from `dark` → `undefined` we can do `setTheme(pendingTheme)` so everything is up-to-date.

trm217 replied:
I added some more utilities to test storage-events. I'll implement the logic to handle this edge case and write the test later today.

trm217 commented:
Allrighty so I did the following:
- add a new state to track the values that were given to the storage-event, while forcedTheme was set (the last called value is always stored.)
- once forcedTheme is unset set theme equal to the value of the last storage-event (if there was one)
- cleanup tests

trm217 commented:
Crap, didn't realise that this PR was still not merged... The forks is deleted and the changes are lost 😒
